Add tests for prediction pipeline; add keras model adapter

constantinpape opened the pull request:
I have started to add the keras model adapter (not tested yet) and tests for the prediction pipeline.
One thing I noticed is that the device management in the prediction pipeline seems to be tailored to torch right now.
We should think about more framework agnostic device management.

Also, the test for the torchscript model currently fails with `E           RuntimeError: [enforce fail at inline_container.cc:145] . PytorchStreamReader failed reading zip archive: unsupported multidisk archive   `.
It seems like the torchscript weights in the example model are corrupted....

constantinpape commented:
@FynnBe the tf / keras tests fail with 

```
self = <bioimageio.spec.shared.utils.UriNodeTransformer object at 0x7f843fff2580>, leaf = PosixPath('/tmp/bioimageio_cache/auto-convert/-821963116692508699/auto_converted_documentation.md')                                                          
                                                                                                                                                                                                                                                       
    def _transform_Path(self, leaf: pathlib.Path):                                                                                                                                                                                                     
>       assert not leaf.is_absolute()                                                                                                                                                                                                                  
E       AssertionError 
```

So again some issue with the auto converted documentation...

FynnBe commented:
> @FynnBe the tf / keras tests fail with
> 
> ```
> self = <bioimageio.spec.shared.utils.UriNodeTransformer object at 0x7f843fff2580>, leaf = PosixPath('/tmp/bioimageio_cache/auto-convert/-821963116692508699/auto_converted_documentation.md')                                                          
>                                                                                                                                                                                                                                                        
>     def _transform_Path(self, leaf: pathlib.Path):                                                                                                                                                                                                     
> >       assert not leaf.is_absolute()                                                                                                                                                                                                                  
> E       AssertionError 
> ```
> 
> So again some issue with the auto converted documentation...
> 
> p.s. this is also happening for all the example zero-cost models we want to support. Is there a way to properly fix this @FynnBe?
> Otherwise I would strongly suggest that we drop the requirement for the local doc again, this is a huge pain!

should be fixed by https://github.com/bioimage-io/spec-bioimage-io/pull/218 
Can't reproduce atm (missing keras env...)

constantinpape commented:
Yes, that fixes the particular error, now it fails with 
```
self = Model(format_version='0.3.2', name='Small Extracellular Vesicle TEM Segmentation (Fully Residual U-Net)', type='model'...e/Small_Extracellular_Vesicle_TEM_Segmentation_(Fully_R
esidual_U-Net)p0_unzipped/tensorflow_saved_model_bundle.zip'))})                                                                                                                       
                                                                                                                                                                                       
    def __post_init__(self):                                                                                                                                                           
        from . import nodes, raw_nodes                                                                                                                                                 
                                                                                                                                                                                       
        if not isinstance(self, (nodes.Node, raw_nodes.RawNode)):                                                                                                                      
            raise TypeError("base nodes should not be instantiated!")                                                                                                                  
                                                                                                                                                                                       
        for f in dataclasses.fields(self):                                                                                                                                             
            if getattr(self, f.name) is missing and (                                                                                                                                  
                get_origin(f.type) is not Union or not isinstance(missing, get_args(f.type))                                                                                           
            ):                                                                                                                                                                         
>               raise TypeError(f"{self.__class__}.__init__() missing required argument: '{f.name}'")                                                                                  
E               TypeError: <class 'bioimageio.spec.model.v0_3.nodes.Model'>.__init__() missing required argument: 'source' 
```

FynnBe commented:
Yep. It's a missing "Union[Missing, ..." annotation in the raw model node. I'll fix it when I'm back later today

FynnBe commented:
Current error:
```
from bioimageio.core.utils import get_nn_instance
E   ModuleNotFoundError: No module named 'bioimageio.core.utils
```
get_nn_instance was moved to the pytorch model adapter...

constantinpape commented:
@FynnBe @k-dominik I have added the keras model adapter, added tests for the prediction pipeline and moved the device logic into the model adapters so that we don't rely on torch for this.

The tests for tf and keras are still failing because of issues with loading the weights; but I think we should go ahead and merge this anyway, because I think these issues are model specific and will need to be fixed together with @esgomezm.

k-dominik commented on bioimageio/core/prediction_pipeline/_model_adapters/_onnx_model_adapter.py:
```diff
@@ -30,6 +30,7 @@ def __init__(self, *, bioimageio_model: nodes.Model, devices=List[str]):
         onnx_inputs = self._session.get_inputs()
         assert len(onnx_inputs) == 1, f"expected onnx model to have one input got {len(onnx_inputs)}"
         self._input_name = onnx_inputs[0].name
+        # TODO onnx device management
```
what happens with empty devices? Does it run on cpu? In any case maybe one should issue a warning here. (same goes for `TensorflowModelAdapterBase`).

constantinpape replied:
The empty device list is just a place holder until someone implements proper device management here... I think in both cases (tf and onnx) it runs on CPU. 
I agree, we should at least add a warning that arguments are ignored (if the devices are not None).

constantinpape replied:
I added warnings for this.

k-dominik commented on tests/conftest.py:
```diff
@@ -2,6 +2,28 @@
 
+# set 'skip_<FRAMEWORK>' flags as global pytest variables,
+# to deselect tests that require frameworks not available in current env
+def pytest_configure():
```
initially I thought "awesome", but now, I'm not sure if this is a good idea. First of all creating the development environment would ideally install all dependencies (imo) to run all tests. Then skipping some for missing deps is not necessary. The only issue might arise if there are conflicts that make it impossible to install everything into the same env.

constantinpape replied:
It also depends if we want tests to run for users who might only set up an environment with a subset of the dependencies.
I think it would also be quite useful to add another "blank" env for testing that does not contain any of the frameworks, to check that the module can be imported, and that base functionality works, with missing frameworks.

FynnBe reviewed (approved):
some minor comments only

constantinpape commented on bioimageio/core/predict.py:
```diff
@@ -30,7 +30,8 @@ def main():
     input_tensors = [np.load(ipt) for ipt in args.images]
-    tagged_data = [xr.DataArray(ipt_tensor, dims=tuple(ipt.axes)) for ipt_tensor, ipt in zip(input_tensors, model.inputs)]
+    tagged_data = [xr.DataArray(ipt_tensor, dims=tuple(ipt.axes))
+                   for ipt_tensor, ipt in zip(input_tensors, model.inputs)]
```
I think the line was longer than 120.

constantinpape commented on tests/weight_converter/torch/test_onnx.py:
```diff
@@ -1,23 +1,15 @@
-@pytest.mark.skipif(rt is None, reason="requires onnx")
+@pytest.mark.skipif(pytest.skip_torch, reason="requires pytorch")
 def test_torchscript_converter(unet2d_nuclei_broad_model, tmp_path):
     from bioimageio.core.weight_converter.torch.onnx import convert_weights_to_onnx
```
No, the weight converter does not require onnx. It only has it as optional dependency to check the result, see https://github.com/bioimage-io/python-bioimage-io/blob/main/bioimageio/core/weight_converter/torch/onnx.py#L8-L11.

constantinpape commented:
I will go ahead and merge this.
